forem/forem
# Made published_at immutable after an article has been published once

## What type of PR is this? (check all applicable)

## Description
We don't want to allow users to use unpublishing and republishing as a hack, and bump an old post to have a current published at [Slack thread](https://forem-team.slack.com/archives/C7GE84DEJ/p1660845549633789?thread_ts=1660841521.696969&cid=C7GE84DEJ)
In this pr I:
- removed bumping published_at to current on republishing
- rewrote validation for `published_at` on update, currently it:
  - allows future or current published_at when publishing for the first time (from a draft)
  - allows any published_at for articles that were `published_from_feed` when publishing for the first time (from a draft)
  - doesn't allow changing `published_at` for articles that have already been published (had `published_at` already)
- added more specs for unpublishing, publishing, and republishing, including the tests for the frontmatter

## Related Tickets & Documents
- Related Issue #15858 

## QA Instructions, Screenshots, Recordings
- enable the feature flag `FeatureFlag.enable(:schedule_articles)`
Try from both basic and rich editors:
- publish an article from a draft for the first time => published_at should be set to current time
- publsh an article, unpublish, then republish => published_at should remain as it was on the first publication
- unpublish a published article => published_at should remain as it was
- try changing `published_at` on a published_article => it should not be allowed
- try changing `published_at` of a scheduled article => it shoud be allowed to change it to the future, but not to the past

## Added/updated tests?


## [Forem core team only] How will this change be communicated?
